File: cpp_validator/src/cubic_spline.cpp
#include "cubic_spline.hpp"
#include <stdexcept>

CubicSpline::CubicSpline(
    const std::vector<float>& x_knots,
    const std::vector<float>& a_coef,
    const std::vector<float>& b_coef,
    const std::vector<float>& c_coef,
    const std::vector<float>& d_coef,
    std::shared_ptr<ONNXModel> model
)
    : x_knots_(x_knots)
    , a_coef_(a_coef)
    , b_coef_(b_coef)
    , c_coef_(c_coef)
    , d_coef_(d_coef)
    , model_(std::move(model))
{
    // Validate inputs
    if (x_knots.empty()) {
        throw std::invalid_argument("x_knots cannot be empty");
    }

    size_t n = x_knots.size();
    if (a_coef.size() != n || b_coef.size() != n || c_coef.size() != n || d_coef.size() != n) {
        throw std::invalid_argument("Coefficient arrays must have same size as knots array");
    }

    if (!model_) {
        throw std::invalid_argument("Model cannot be null");
    }
}

std::vector<float> CubicSpline::evaluate(const std::vector<float>& x) const {
    std::vector<std::pair<std::string, std::vector<float>>> inputs = {
        {"x", x},
        {"x_known", x_knots_},
        {"a_coef", a_coef_},
        {"b_coef", b_coef_},
        {"c_coef", c_coef_},
        {"d_coef", d_coef_}
    };

    return model_->run(inputs, "y");
}

File: java_validator/src/main/java/com/splinevalidator/CubicSpline.java
package com.splinevalidator;

import org.apache.commons.math3.analysis.interpolation.SplineInterpolator;
import org.apache.commons.math3.analysis.polynomials.PolynomialSplineFunction;

public class CubicSpline {
    private final double[] xKnots;
    private final double[] aCoef;
    private final double[] bCoef;
    private final double[] cCoef;
    private final double[] dCoef;
    private final ONNXModel model;

    public CubicSpline(double[] xKnots, double[] aCoef, double[] bCoef, double[] cCoef, double[] dCoef, ONNXModel model) {
        if (xKnots == null || xKnots.length == 0) {
            throw new IllegalArgumentException("xKnots cannot be null or empty");
        }

        int n = xKnots.length;
        if (aCoef.length != n || bCoef.length != n || cCoef.length != n || dCoef.length != n) {
            throw new IllegalArgumentException("Coefficient arrays must have same size as knots array");
        }

        if (model == null) {
            throw new IllegalArgumentException("Model cannot be null");
        }

        this.xKnots = xKnots.clone();
        this.aCoef = aCoef.clone();
        this.bCoef = bCoef.clone();
        this.cCoef = cCoef.clone();
        this.dCoef = dCoef.clone();
        this.model = model;
    }

    public float[] evaluate(float[] x) {
        // Convert inputs to float arrays for ONNX
        float[] xKnotsFloat = new float[xKnots.length];
        float[] aCoefFloat = new float[aCoef.length];
        float[] bCoefFloat = new float[bCoef.length];
        float[] cCoefFloat = new float[cCoef.length];
        float[] dCoefFloat = new float[dCoef.length];

        // Convert all inputs to float and ensure proper ordering
        for (int i = 0; i < xKnots.length; i++) {
            xKnotsFloat[i] = (float) xKnots[i];
            aCoefFloat[i] = (float) aCoef[i];
            bCoefFloat[i] = (float) bCoef[i];
            cCoefFloat[i] = (float) cCoef[i];
            dCoefFloat[i] = (float) dCoef[i];
        }

        // Run the model with preprocessed inputs
        return model.run(x, xKnotsFloat, aCoefFloat, bCoefFloat, cCoefFloat, dCoefFloat);
    }
}
